### DISCUSSION TITLE
Example of custom field implementation, that stores data as object or array

### DISCUSSION THREAD

**Discussion by MurzNN** [keystonejs/keystone#5318] [Feature requests]
Before #195 will be implemented, will be good to have some example of such field, that stores his data in supportable databases (MongoDB or PostgreSQL JSONB field) not as simple JSON text string, but as structured object.

We already have the `MultiCheck` field from `examples/custom-fields` that must do this thing, but in database it store data as simple string, here is example of MongoDB document:
```json
{
    "_id": "5fd9e6101dab3e967f37364c",
    "name": "Super Movie",
    "rating": 4,
    "categories": "{\"Action\":true,\"Comedy\":true,\"Drama\":false}",
    "__v": 0
}
```
The first easy problem with current storage type is that I can't filter list of movies by category, for example to select all movies with "Comedy" category. Without this feature - categories field becomes useless.
So will be good to store same data as array, like this:
```json
{
    "_id": "5fd9e6101dab3e967f37364c",
    "name": "Super Movie",
    "rating": 4,
    "categories": [
        "Action",
        "Comedy "
    ],
    "__v": 0
}
```
or via object like this:
```json
{
    "_id": "5fd9e6101dab3e967f37364c",
    "name": "Super Movie",
    "rating": 4,
    "categories": {
        "Action": true,
        "Comedy": true,
        "Drama": false
    },
    "__v": 0
}
```

Also we already have `examples-next/basic` example, that try to implement this via `fields-document` field type, but this is too complex field in implementation.

So, can you extend current `MultiCheck` field in `examples-next` to implement array or object storage type (at first - maybe with easy Admin UI showing editable json in textarea), or this is not possible yet?

**Comment by MurzNN** [comment]
Here is example of how the `fields-document` field store the data in MongoDB document:
```
{
    "_id": "5fd9ede39085a0c163ae7678",
    "title": "Test document field",
    "content": {
        "document": [
            {
                "type": "paragraph",
                "children": [
                    {
                        "text": "Bold text",
                        "bold": true
                    }
                ]
            },
            {
                "type": "paragraph",
                "children": [
                    {
                        "text": " and regular text"
                    }
                ]
            }
        ]
    },
    "author": "5fd9ed8f9085a0c163ae7676",
    "__v": 0
}
```

**Comment by pahaz** [comment]
@MurzNN hi! We have a JSON field for such cases: https://github.com/8iq/nodejs-hackathon-boilerplate-starter-kit/tree/51a7c37/apps/_back02keystone/custom-fields/Json

And Some tests for different types of value (Object, Array, String): 
 - store value: https://github.com/8iq/nodejs-hackathon-boilerplate-starter-kit/blob/51a7c37/apps/_back02keystone/_api_tests/_example.test.js#L30-L96
 - filter by value: https://github.com/8iq/nodejs-hackathon-boilerplate-starter-kit/blob/51a7c37/apps/_back02keystone/_api_tests/_example.test.js#L98-L192

This field is working a little differently for mongo and postgres. For example for MongoDB, the value`{}` and `null` are equal.

Schema examples: 
 1) https://github.com/8iq/nodejs-hackathon-boilerplate-starter-kit/blob/51a7c3766ad021fb47c13be615b38ea36fb1a1c4/apps/_back02keystone/schema/User.js#L22
 2) https://github.com/8iq/nodejs-hackathon-boilerplate-starter-kit/blob/51a7c3766ad021fb47c13be615b38ea36fb1a1c4/apps/_back02keystone/schema/Test.js#L105

**Comment by MurzNN** [comment]
@pahaz, thanks for examples! But JSON field is stored in database as one single string, as I understand?

We need to store data in semantic way (as Object in MongoDB, and as separate table with relation to entity in SQL), for make filtering queries by individual JSON field values.

**Reply by pahaz** [reply]
No for mongo it's an `Object` and for Postgres it's `jsonb`.

**Reply by MurzNN** [reply]
Wow, that's great, thanks!

**Reply by pahaz** [reply]
And yes, currently we have base filter support for `_in`, `_not_in`, `_equeal`, `_not_equal` cases. In the future, we will add more cases.

**Reply by pahaz** [reply]
If you want some features you can create an issue. And  🌟 the project :)
